Implement loss scoring for #628

asberk opened the pull request:
This addresses issue #628 . Tests could probably be more complete. @BenjaminBossan , please let me know if/how they could be fleshed out to be more comprehensive. 

As a side note, I have limited familiarity with `PassThroughScoring`, but it seemed, in a way, like overkill. This could be because I'm not properly understanding some edge cases. Anyway, at the time, I thought it was sensible to write an accumulator class that handles the reduction of the batch-level statistics for `criterion`. In case this isn't ideal, is there an easy way to convert the current code to something that uses Skorch tools (*e.g.,* `PassThroughScoring`)?

BenjaminBossan commented:
Thanks a lot @asberk for the PR.

I haven't taken a detailed look at your code yet but two things came to my mind:

1. The docstring of `loss_scoring` needs some clean up, I think it should stick very close to other similar functions. Also, it would be nice to document how to use that function to add a `score` method to the net, similar to the discussion. That could be added to the docstring of `loss_scoring` as well, or maybe in the FAQ section of the docs.

2. I believe from a first glance that the accumulator class is overkill. Shouldn't it be possible to capture the same functionality without a class? The `loss_scoring` could collect the losses and the batch sizes, then pass them to an accumulation function, along with the criterion. Inside that function, you could call `np.average`, similar to how `PassthroughScoring` does it.

I don't think there is a need to use `PassthroughScoring` or to try to share any code with `PassthroughScoring`, the code duplication should be minimal. What do you think about this?

Overall, this looks very good though, nice unit tests.

asberk commented:
Thank you for your comments, @BenjaminBossan . 

1. I've fixed up the docstring and added an example. The example is not an exemplar as `GridSearchCV` looks for higher scores I believe. If you suggest an avenue that more simply showcases the behaviour, I'll update it to that.
2. I've removed the accumulator abstraction. There is now also a test to make sure `loss_scoring` computes the same value as `net.criterion_(y_proba, labels)`.

BenjaminBossan commented on skorch/utils.py:
```diff
@@ -544,6 +544,7 @@ class TeeGenerator:
     generator more than once.
 
     """
+
```
Could you please undo this change? I think it was accidentally checked in.

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+    ----------
+    net : skorch.NeuralNet
+        A fitted Skorch :class:`.NeuralNet` object.
+    X : input data, compatible with skorch.dataset.Dataset
```
Could you please add an empty line between parameters?

BenjaminBossan commented on skorch/tests/test_scoring.py:
```diff
@@ -0,0 +1,91 @@
+        return classifier_data
+
+    @pytest.fixture(scope="module")
+    def val_data(self, classifier_data):
```
I think for testing this function, we don't actually need separate validation data, we can just use the training data.

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def loss_scoring(net: NeuralNet, X, y=None):
+    """
```
```suggestion
    """Calculate score using the criterion of the net
    
    Use the exact same logic as during model training to calculate the score.
```

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+
+def loss_scoring(net: NeuralNet, X, y=None):
+    """
+    loss_scoring(net, X, y=None)
+
+    Computes the loss of ``net`` on data (``X``, ``y``).
```
```suggestion
```

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+    combining skorch models with sklearn objects that rely on the model's
+    ``score`` method. Below is an example using GridSearchCV.
+
+    >>> class ScoredNet(skorch.NeuralNetClassifier):
```
Thanks for adding the elaborate example. I think, however, that it could be simplified. 

1) You could just show how to subclass the net here.
2) You could go with the grid search example, but in that case you actually don't need to subclass the net. You can pass the `scoring=loss_scoring` to `GridSearchCV` and it should result in the same outcome.

Assume that you go with the second option: I would then add the subclassing example to the docs, probably to the FAQ ("How to implement a score method on the net that returns the loss?" or something along these lines).

asberk replied:
I wound up with a combination of 1. and 2. The docstring now shows the sub-classing. The FAQ now has an example of `loss_scoring(net, X, y)` as well as an example using `GridSearchCV`.

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+    ``score`` method. Below is an example using GridSearchCV.
+
+    >>> class ScoredNet(skorch.NeuralNetClassifier):
+    ...     def score(self, X, y=None, lower_is_better=False):
```
```suggestion
    ...     def score(self, X, y=None):
```

I would remove this because it doesn't conform to the sklearn API. Also, if `lower_is_better = True`, which is always true for PyTorch criteria, we should return `-loss_value` and not vice versa? Otherwise, something like grid search would choose the model with the _highest_ loss.

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,90 @@
+        this function returns a ``np.ndarray`` object.
+
+    """
+    net.check_is_fitted()
```
Good attention to detail

asberk commented:
I believe I have addressed the changes you've mentioned, @BenjaminBossan . Thank you for the detailed list and comments!

Funny enough, there were tests for `reduction` ∈ {`'mean'` `'sum'`} in that class that no longer exists, and I forgot to transfer that over. The `reduction` fixture uses `'mean'` and `'sum'` only, and `test_scored_net_matches_criterion_value` tests score using the same `reduction` that was used to train the network. This could be modified if you think it is insufficient: so that a trained net's scoring is tested for both `'mean'` and `'sum'`. 

I made a separate test for when the reduction is `'none'` since the behaviour is so different, and since the reduction must be altered after fitting. There is also a test asserting that an unknown reduction raises.

BenjaminBossan commented on docs/user/FAQ.rst:
```diff
@@ -357,3 +357,50 @@ layer, and that the number of features can be determined by
+-------------------------------------------------------------------
+
+Sometimes, it is useful to be able to compute the loss of a net from within
+``skorch`` (*e.g.,* when a net is part of an ``sklearn`` pipeline). The function
```
```suggestion
``skorch`` (e.g. when a net is part of an ``sklearn`` pipeline). The function
```
Or why should "e.g." be in italics?

asberk replied:
I'd italicized it because it's Latin, not English. But it does not matter on the whole, and I admit it is better to opt for consistency. I suppose better readability of the plaintext is a bonus of no italics.

BenjaminBossan replied:
Okay, I didn't know that. As you said, for consistency (and since it's easier to type), let's do without.

BenjaminBossan commented on docs/user/FAQ.rst:
```diff
@@ -357,3 +357,50 @@ layer, and that the number of features can be determined by
+    net = skorch.NeuralNetClassifier(module).fit(X, y)
+    print(loss_scoring(net, X, y))
+
+The second example shows how to sub-class :class:`skorch.NeuralNetClassifier` to
```
```suggestion
The second example shows how to sub-class :class:`skorch.classifier.NeuralNetClassifier` to
```

At least for me that was necessary to make the link work.

BenjaminBossan commented on skorch/scoring.py:
```diff
@@ -0,0 +1,82 @@
+    iterator = net.get_iterator(dataset, training=False)
+    history = {"loss": [], "batch_size": []}
+    reduction = net.criterion_.reduction
+    assert reduction in ["mean", "sum", "none",], (
```
Could you please raise a `ValueError` instead?

BenjaminBossan commented on skorch/tests/test_scoring.py:
```diff
@@ -0,0 +1,114 @@
+        ).fit(X, y)
+        net.set_params(criterion__reduction="none")
+        output = net.score(X, y)
+        assert output.shape[0] == X.shape[0]
```
I believe this test can be made a little bit more sophisticated than just checking the shape. If you first calculate the loss with `reduction='sum'` and then with `'none'`, the sum of the latter should equal the former.

asberk replied:
Good idea. I kept the `assert` for `shape` equality, and added the extra lines for asserting value equality.

BenjaminBossan commented:
Thanks @asberk for this feature, you did excellent work. I'm sure many users will find it useful.

asberk commented:
Thank you @BenjaminBossan for the learning opportunities. Happy to contribute.
